bindings/rust/evmc-vm/src/lib.rs

```diff
@@ -61,6 +69,170 @@ impl ExecutionResult {
+            )
+        }
+    }
+
+    pub fn get_code_size(&mut self, addr: &ffi::evmc_address) -> usize {
+        unsafe {
+            assert!((*self.context.host).get_code_size.is_some());
+            (*self.context.host).get_code_size.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+            )
+        }
+    }
+
+    pub fn get_code_hash(&mut self, addr: &ffi::evmc_address) -> ffi::evmc_bytes32 {
+        unsafe {
+            assert!((*self.context.host).get_code_size.is_some());
+            (*self.context.host).get_code_hash.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+            )
+        }
+    }
+
+    pub fn copy_code(
+        &mut self,
+        addr: &ffi::evmc_address,
+        code_offset: usize,
+        buffer: &mut [u8],
+    ) -> usize {
+        unsafe {
+            assert!((*self.context.host).copy_code.is_some());
+            (*self.context.host).copy_code.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+                code_offset,
+                buffer.as_mut_ptr(),
+                buffer.len(),
+            )
+        }
+    }
+
+    pub fn selfdestruct(&mut self, addr: &ffi::evmc_address, beneficiary: &ffi::evmc_address) {
+        unsafe {
+            assert!((*self.context.host).selfdestruct.is_some());
+            (*self.context.host).selfdestruct.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+                beneficiary as *const ffi::evmc_address,
+            )
+        }
+    }
+
+    pub fn call(&mut self, message: &ffi::evmc_message) -> ExecutionResult {
+        unsafe {
+            assert!((*self.context.host).call.is_some());
+            (*self.context.host).call.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                message as *const ffi::evmc_message,
+            )
+            .into()
+        }
+    }
+
+    pub fn get_block_hash(&mut self, num: i64) -> ffi::evmc_bytes32 {
+        unsafe {
+            assert!((*self.context.host).get_block_hash.is_some());
+            (*self.context.host).get_block_hash.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                num,
+            )
+        }
+    }
+
+    pub fn emit_log(
+        &mut self,
+        addr: &ffi::evmc_address,
+        data: &[u8],
+        topics: &[ffi::evmc_bytes32],
+    ) {
+        unsafe {
+            assert!((*self.context.host).emit_log.is_some());
+            (*self.context.host).emit_log.unwrap()(
+                self.context as *mut ffi::evmc_context,
+                addr as *const ffi::evmc_address,
+                data as *const [u8] as *const u8,
+                data.len(),
+                topics as *const [ffi::evmc_bytes32] as *const ffi::evmc_bytes32,
```

Comment:
Not sure about alignment here and this looks nasty. At the bare minimum with need a FIXME note and explanation.